### ISSUE TITLE
meta: ensure that only miller can push to 'master' branch

### ISSUE THREAD

**Issue by umlaeute** [pure-data/pure-data#2138]
### TL;DR

i've add some more branch-protection rules, trying to ensure that only @millerpuckette can push to the `master` branch.

as it is hard to test, this might have accidentally locked out everybody (including @millerpuckette) from pushing to the `master` branch.

so if there's a problem the next time miller is trying to push to `master`, please ping me, so I can undo the branch protection.

## longer and duller explanation

This is a followup on https://github.com/pure-data/pure-data/pull/2137#issuecomment-1818997974 (where i accidentally merged a PR into `master` instead of the intended `develop` branch).

#### prerequisites
1. following our "benevolent dictator for life" development model, the **only** user with write access to the `master` branch should be @millerpuckette 
2. to ensure a slightly easier workflow (and integration with the iem-ci), we also have a number of contributors who can directly push to this repository
3. general github project management is done by multiple people (well, mostly it's me, but the point is, that it is *not* miller), so they require `Administrator` role.

these givens should be enforced on the server-side, so that it is really hard to accidentally push to the `master` branch, even for people (such as myself, who are *Administrator* in this repo)

#### implementation

github allows for *branch protection* rules, which include things like "PR before merging" or "Restrict who can push to branch".
The latter rule is exactly what we want, but unfortunately it comes with a caveat:
> #### People, teams, or apps with push access
> Organization administrators, repository administrators, and users with the Maintain role.
> (Admins can always push. Users with the Maintain role can push when required status checks pass.)

so it obviously does not help against rogue Admins.

There is also an option that says
> #### Do not allow bypassing the above settings 
> The above settings will apply to administrators and custom roles with the "bypass branch protections" permission.

but of course, the catch is, that the "People with push access" setting comes *below* this setting, so it doesn't prevent Admins from pushing.

Luckily there's also an option (before the "do not allow bypassing" setting that says:

> #### Lock branch
>  Branch is read-only. Users cannot push to the branch.

so the current setup is:
- [x] Lock branch (make `master` read-only)
- [x] Do not allow bypassing the above settings (so *Admins* cannot push to the read-only branch)
- [x] Grant push access to a  single user (so @millerpuckette can still push to `master`)

hopefully this works as intended.

**Comment by umlaeute** [comment]
i've at least tested whether I can (force) push to the `master` branch, and it seems that I'm no longer allowed to do so:

```
$ git push -f
Total 0 (delta 0), reused 0 (delta 0), pack-reused 0
remote: error: GH006: Protected branch update failed for refs/heads/master.
remote: error: Cannot change this locked branch
To github.com:pure-data/pure-data.git
 ! [remote rejected]   master -> master (protected branch hook declined)
error: failed to push some refs to 'github.com:pure-data/pure-data.git'
```

so that part appears to be working as intended.

**Comment by millerpuckette** [comment]
I don't ever use it, thanks :)

**Comment by umlaeute** [comment]
hmm.
i've done another test (with me having an exception to the "Read-only branch" rule (like miller), and it seems i can no longer `push` nor `force-push`.

Which means, that this experiment did not do what I hoped it was doing.

I have therefore reverted the Branch protection rules to exclude the ~~`Lock branch (read-only branch)`~~ rule

**Comment by umlaeute** [comment]
@millerpuckette thanks for the quick reply neverthless :-)
